[WebGPU] RemoteShaderModule::CompilationInfo results in an IPC exception

mwyrzykowski opened the pull request:
#### bfcd800c0db8421512486ba84867bc833e9be209
<pre>
[WebGPU] RemoteShaderModule::CompilationInfo results in an IPC exception
<a href="https://bugs.webkit.org/show_bug.cgi?id=252586">https://bugs.webkit.org/show_bug.cgi?id=252586</a>
&lt;radar://105688260&gt;

Reviewed by Myles C. Maxfield.

This is an async reply, so use sendWithAsyncReply as async reply from a
sync IPC is not supported.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
(WebKit::WebGPU::RemoteShaderModuleProxy::compilationInfo):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.h:

Canonical link: <a href="https://commits.webkit.org/260664@main">https://commits.webkit.org/260664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a11b564bf450a47e2fc4ae41db80918b81e80b2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17822 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41570 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/36 "Updated wpe dependencies (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118021 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9094 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100959 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114484 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14454 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97692 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42436 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96428 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84281 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10620 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30681 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11382 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7594 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16769 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50271 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7364 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12968 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
```diff
@@ -48,13 +48,14 @@ RemoteShaderModuleProxy::~RemoteShaderModuleProxy()
-        return PAL::WebGPU::CompilationMessage::create(WTFMove(compilationMessage.message), compilationMessage.type, compilationMessage.lineNum, compilationMessage.linePos, compilationMessage.offset, compilationMessage.length);
+    auto sendResult = sendWithAsyncReply(Messages::RemoteShaderModule::CompilationInfo(), [callback = WTFMove(callback)](auto messages) mutable {
+        auto backingMessages = messages.map([](CompilationMessage compilationMessage) {
+            return PAL::WebGPU::CompilationMessage::create(WTFMove(compilationMessage.message), compilationMessage.type, compilationMessage.lineNum, compilationMessage.linePos, compilationMessage.offset, compilationMessage.length);
```
Is it possible construct `WebGPU::CompilationMessage` directly? Feels strange to me to unpack a struct to pass to a construct to be packed into a struct.

djg replied:
Unless you've internalised the C++ Koan of the Big Barking Dog. https://twitter.com/tvaneerd/status/977257752820113413

mwyrzykowski replied:
We could pass it directly, but I think the point is to avoid using ::WebGPU::Type inside the PAL layer which is why the unpacking to non-::WebGPU::type is here. As to why there are both the PAL::WebGPU and ::WebGPU seemingly identical types instead of directly using PAL:: types here, I would defer to Myles

djg reviewed (approved):
:shipit:

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.messages.in:
```diff
@@ -24,7 +24,7 @@
 
 messages -> RemoteShaderModule NotRefCounted Stream {
-    void CompilationInfo() -> (Vector<WebKit::WebGPU::CompilationMessage> messages) Synchronous
+    void CompilationInfo() -> (Vector<WebKit::WebGPU::CompilationMessage> messages)
```
Woah, can we just do this?

mwyrzykowski replied:
Yes :), sendWithAsyncReply will complain if the message is marked Synchronous

litherum reviewed (approved):
Looks good.

mwyrzykowski committed:
[WebGPU] RemoteShaderModule::CompilationInfo results in an IPC exception
https://bugs.webkit.org/show_bug.cgi?id=252586
<radar://105688260>

Reviewed by Myles C. Maxfield.

This is an async reply, so use sendWithAsyncReply as async reply from a
sync IPC is not supported.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
(WebKit::WebGPU::RemoteShaderModuleProxy::compilationInfo):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.h:

Canonical link: https://commits.webkit.org/260664@main
